File: examples/threads/PingPong.java

/**
 * A simple thread example that sort of simulates ping pong.
 * 
 * @author amit
 */
class PingPong extends Thread {
	private String word; // what word to print
	private int delay; // how long to pause, in milliseconds

	public PingPong(String whatToSay, int delayTime) {
		word = whatToSay;
		delay = delayTime;
	}

	public void run() {
		try {
			for (;;) {
				System.out.println(word + " ");
				sleep(delay); // wait until next time
			}
		} catch (InterruptedException e) {
			return; // end this thread
		}
	}

	public static void main(String[] args) {
		if (args.length != 1) {
			System.err.println("Usage: java PingPong <delay>");
			System.exit(1);
		}
		int delay = Integer.parseInt(args[0]);
		new PingPong("ping", delay).start();
		new PingPong("PONG", delay).start();
	}
}

File: examples/go/threads/PingPong.go
//Author: Amit Jain & Paul Ellis

package main

import (
	"fmt"
	"time"
	"os"
	"sync"
	"strconv"
)

type PingPong struct {
	word string
	delay int
}

func NewPingPong(word string, delay int) *PingPong{
	p := PingPong{word:word, delay:delay}
	return &p
}

func (p *PingPong)Run(){
	for{
		fmt.Println(p.word + " ")
		d,_ := time.ParseDuration(strconv.Itoa(p.delay)+"ms")
		time.Sleep(d)
	}
}

func main(){
	args := os.Args[1:]
	
	if len(args) != 1{
		fmt.Println("Usage: go run PingPong.go <delay(ms)>")
		os.Exit(1)
	}

	delay,_ := strconv.Atoi(args[0])
	var wg sync.WaitGroup

	wg.Add(1)
	go NewPingPong("ping", delay).Run()
	go NewPingPong("PONG", delay).Run()

	wg.Wait()
}